### PR TITLE
Instant export

### PR DESCRIPTION
**Pull request by thrillfall**
Instant post-workout export to external storage

@dennisguse in a previous review you mentioned reusing ExportActivity. Since this does not exists anymore could you please elaborate on what could be reused now?

### CODE REVIEW

**Comment by dennisguse** [comment]
@thrillfall Is this ready for review?

**Comment by thrillfall** [comment]
> @thrillfall Is this ready for review?

yes

**Comment by thrillfall** [comment]
i tried to in TrackRecording Service
```
if (PreferencesUtils.shouldInstantExportAfterWorkout(getApplicationContext())) {
            TrackFileFormat trackFileFormat = PreferencesUtils.getExportTrackFileFormat(this);

            Uri directoryAsString = PreferencesUtils.getInstantExportDirectoryUri(getApplicationContext());
            exportLastTrackToExternalStorage(directoryAsString, getApplicationContext(), contentProviderUtils, trackFileFormat, this.getRecordingTrackId());
        }
```
but trackRecordingId is null

File: src/main/java/de/dennisguse/opentracks/services/TrackRecordingService.java
Comment (thrillfall) [reply src/main/java/de/dennisguse/opentracks/services/TrackRecordingService.java:197]: i did
```
public static Uri getInstantExportDirectoryUri(Context context)
    {
        String singleExportDirectorySettingsKey = getString(context, R.string.settings_single_export_directory_key, null);
        String directoryAsString = getSharedPreferences(context).getString(singleExportDirectorySettingsKey, "invalid uri");
        return Uri.parse(directoryAsString);
    }
```
but that gives me only `invalid url`. With the previous implementation it worked

Comment (dennisguse) [reply src/main/java/de/dennisguse/opentracks/services/TrackRecordingService.java:197]: Then the key is not set and you get the default value.
Can you check that you are setting the value for this key successfully?
Because it looks good.

**Comment by dennisguse** [comment]
@thrillfall After the recording is stopped, the recordingTrackId in the sharedpreferences is set to invalid, so you cannot get it from there.
You need to get it before the actual stop is conducted.
Take a look at `TrackRecordingServiceConnection.stopRecording()`

But I would recommend to get the trackId from the service rather than the preferences.

**Comment by thrillfall** [comment]
I am calling this.getRecordingTrackId(). From the recording service.

**Comment by dennisguse** [comment]
...but it is *after* stopping, right?
The service also sets the property to null.

File: src/main/java/de/dennisguse/opentracks/io/file/post_workout_export/InstantPostWorkoutExport.java
Comment (thrillfall) [reply src/main/java/de/dennisguse/opentracks/io/file/post_workout_export/InstantPostWorkoutExport.java]: done

File: src/main/java/de/dennisguse/opentracks/io/file/post_workout_export/InstantPostWorkoutExport.java
Comment (thrillfall) [reply src/main/java/de/dennisguse/opentracks/io/file/post_workout_export/InstantPostWorkoutExport.java]: done

File: src/main/java/de/dennisguse/opentracks/services/TrackRecordingService.java
Comment (thrillfall) [reply src/main/java/de/dennisguse/opentracks/services/TrackRecordingService.java]: the second line of code is totally useless (copy and paste error). it works now

File: src/main/java/de/dennisguse/opentracks/io/file/post_workout_export/InstantPostWorkoutExport.java
Comment (thrillfall) [reply src/main/java/de/dennisguse/opentracks/io/file/post_workout_export/InstantPostWorkoutExport.java]: done

File: src/main/java/de/dennisguse/opentracks/io/file/post_workout_export/LastTrackExportThread.java
Comment (thrillfall) [reply src/main/java/de/dennisguse/opentracks/io/file/post_workout_export/LastTrackExportThread.java:12]: done

File: src/main/java/de/dennisguse/opentracks/services/TrackRecordingService.java
Comment (thrillfall) [reply src/main/java/de/dennisguse/opentracks/services/TrackRecordingService.java]: done

File: src/main/res/values/settings.xml
Comment (thrillfall) [reply src/main/res/values/settings.xml:34]: i'll look into that next

Comment (thrillfall) [reply src/main/res/values/settings.xml]: done

**Comment by thrillfall** [comment]
i think i have fixed now all remaining issues. I leave it to @dennisguse to hit the resolve buttons

File: src/main/java/de/dennisguse/opentracks/settings/SettingsActivity.java
Comment (thrillfall) [reply src/main/java/de/dennisguse/opentracks/settings/SettingsActivity.java]: done

File: src/main/java/de/dennisguse/opentracks/io/file/post_workout_export/LastTrackExportThread.java
Comment (thrillfall) [reply src/main/java/de/dennisguse/opentracks/io/file/post_workout_export/LastTrackExportThread.java:12]: if i do it like this
```
  try (OutputStream outputStream = context.getContentResolver().openOutputStream(file.getUri())) {
            new Thread(() -> {
                TrackExporter trackExporter = trackFileFormat.newTrackExporter(context);
                trackExporter.writeTrack(track, outputStream);
            }).start();
            Toast.makeText(context, R.string.toast_message_instant_track_was_exported, Toast.LENGTH_SHORT).show();
        }   catch (FileNotFoundException e) {
            Log.e(TAG, "Unable to open file " + file.getName(), e);
```
i get 
`KmzTrackExporter: Unable to write track
    java.io.IOException: Stream Closed
        at java.io.FileOutputStream.write(FileOutputStream.java:391)
        at java.util.zip.DeflaterOutputStream.deflate(DeflaterOutputStream.java:260)
        at java.util.zip.ZipOutputStream.closeEntry(ZipOutputStream.java:257)
        at de.dennisguse.opentracks.io.file.exporter.KmzTrackExporter.writeTrack(KmzTrackExporter.java:88)
        at de.dennisguse.opentracks.io.file.exporter.KmzTrackExporter.writeTrack(KmzTrackExporter.java:74)
        at de.dennisguse.opentracks.io.file.exporter.InstantPostWorkoutExport.lambda$exportLastTrackToExternalStorage$0(InstantPostWorkoutExport.java:37)
        at de.dennisguse.opentracks.io.file.exporter.-$$Lambda$InstantPostWorkoutExport$xGSFr8OIF-hicfvvOJgeo7zK29s.run(Unknown Source:8)
        at java.lang.Thread.run(Thread.java:923)
E/KmzTrackExporter: Unable to close zip input stream`

**Comment by thrillfall** [comment]
@dennisguse I cannot fix the "thread as lambda function" issue (https://github.com/OpenTracksApp/OpenTracks/pull/383#discussion_r476662156) by myself. So this PR is as far is i got from my side.

File: src/main/java/de/dennisguse/opentracks/io/file/post_workout_export/LastTrackExportThread.java
Comment (pstorch) [reply src/main/java/de/dennisguse/opentracks/io/file/post_workout_export/LastTrackExportThread.java:12]: @thrillfall you have to put the try .. catch inside the Thread

**Comment by dennisguse** [comment]
@thrillfall I was off for the weekend.
I guess, I have time during the week to check this PR.

**Comment by dennisguse** [comment]
This code works.
IMHO it does not (yet) integrate nicely with the existing "export all"-feature.
Coding style looks good (some more log lines would nice, but whom am I to judge...).
TL;DR 

Now the stuff that needs to be reworked:
1. UX issue regarding settings: we have now a specific directory for post-recording exports and for export all, we always ask for a directory.
   I propose a "export directory setting"; if enabled export all and post-recording export into the same folder.
1. Simple improvement: post-recording export can only be activated if "export directory setting" is set.
1. Simple improvement: "export directory setting" should show the actual value as preference summary (I guess using useSimpleSummaryProvider)
1. Refactoring required: We already have logic to export multiple tracks: ExportProgressDialogFragment.ExportThread
    InstantPostWorkoutExport and ExportThread should use the same code to export (ExportUtils?)
    Also the generated filename should be same for both.
1. No specific directory selector activity: reuse DirectoryChooserActivity.
    Just pass a parameter to denote, if the selected URL should be persisted permanently.
    Or rather let SettingsActivity start DirectoryChooserActivity and SettingsActivity can persist the permissions (and also take care of the actual value).

@thrillfall Sorry, it took some time to for me to review this PR.
It just got too big to review in 20mins, so I had to allocate time.

As next step I would recommend the following: can you submit two new PRs?
1. a PR that implements "export directory setting"
2. a PR that implements post-recording exports (as a follow-up)

PS: You can also squash the commits that are just small corrections.
Later on it is easier, the commit history of the whole project is far easier to read.

**Comment by thrillfall** [comment]
@dennisguse thanks for the review. I addressed the following issues:
* Simple improvement: post-recording export can only be activated if "export directory setting" is set.
* Simple improvement: "export directory setting" should show the actual value as preference summary 

i did not address the other suggestions since they are outside of the feature "instant export" that is supposed to be added with this PR.
Having the current implementation of import and export make use of the newly introduced instant export directory settings can be done at a later point (read: PR) from anyone else thus not "holding up" this change set.

This was originally designed in a way not to change any of the existing behavior.

**Comment by dennisguse** [comment]
I don't agree.

We have already enough technical debt in the system that takes a lot of effort - adding more as a shortcut is not an option.
The duplication of code in this case is really not helpful in the long run - removing/merging/fixing it takes way too much resources (and as nobody pays us, it is spare time - even more precious).
Without the changes to unify export all and post-workout export, I cannot merge this PR.

@pstorch @rgmf Just to clarify - am I unreasonable here?

**Comment by rgmf** [comment]
> @pstorch @rgmf Just to clarify - am I unreasonable here?

No, I think not.

----
#### My humble opinion:

> 1. UX issue regarding settings: we have now a specific directory for post-recording exports and for export all, we always ask for a directory.

Yes, please. If I understand well, if user has a directory selected then when you "Export all" then OpenTracks doesn't ask where to export, right? Sound good!!

> 2. Simple improvement: post-recording export can only be activated if "export directory setting" is set.

The first time you click on "Set instant post-workout export directory" and select a directory you can see "not set yet". So you need to exit from Settings and enter once again to see the directory selected and "Instant post-workout export" option.

> 4. Refactoring required: We already have logic to export multiple tracks: ExportProgressDialogFragment.ExportThread
InstantPostWorkoutExport and ExportThread should use the same code to export (ExportUtils?)

This would be great.

Would it possible to use ExportProgressDialogFragment to export last track instead of InstantPostWorkoutExport? When you finish a record then you'll see the dialog exporting the track just recorded... What do you think? Can this disturb the user?

>  Also the generated filename should be same for both.

So important this. In fact, I think we have pending an Issue about the exported name.

----

@thrillfall don't give up... it'll be great to see this feature in OpenTracks and the little I read from your code seem really good. You are about finish it ;)

Maybe what @dennisguse pointed out can help to speed up all this:

> As next step I would recommend the following: can you submit two new PRs?
> 1. a PR that implements "export directory setting"
> 2. a PR that implements post-recording exports (as a follow-up)

**Comment by thrillfall** [comment]
> I don't agree.
> 
> We have already enough technical debt in the system that takes a lot of effort - adding more as a shortcut is not an option.
> The duplication of code in this case is really not helpful in the long run - removing/merging/fixing it takes way too much resources (and as nobody pays us, it is spare time - even more precious).
> Without the changes to unify export all and post-workout export, I cannot merge this PR.
> 
> @pstorch @rgmf Just to clarify - am I unreasonable here?

I removed the code duplication of creating the DocumentFile for export.

Making the "instant export" use the Fragment implementation would mean to reuse a construct of of which most parts are not needed: the fragment, the progress dialog, the iteration the database.
The only business logic thing they share is the creation of the DocumentFile which is unified (https://github.com/OpenTracksApp/OpenTracks/pull/427) and reused (so no more duplication there). 
Any code that is still redundant (creating output stream and starting a thread running the trackwriter) is that because instant export is a different strategy for export. And different strategies should go into different implementations (read: classes) and not be "switched" in a single big class (ExportProgressDialogFragment if i understood you correctly)

I cannot extract and reuse the ExportThread because it depends on the fragment. Instant-export does not need fragment.

Making enabling ExportProgressDialogFragment to use the new export directory setting as a prerequisite to contribute instant-export functionality to the app is something i don't understand.

**Comment by thrillfall** [comment]
> The first time you click on "Set instant post-workout export directory" and select a directory you can see "not set yet". So you need to exit from Settings and enter once again to see the directory selected and "Instant post-workout export" option.

@pstorch 
i see that the the summary provider is being called after i scroll up and back down to the setting and the summary is updated correctly. 
How can i force it to be called (or redraw?) after the SingleTrackExportDirectoryChooserActivity has finished?

**Comment by thrillfall** [comment]
As requested the export all functionality now uses the export directory from settings if it is set. Otherwise the behavior is as before.

What else is needed?

**Comment by dennisguse** [comment]
@thrillfall I will take a look tomorrow. Just the usual request: can you rebase everything onto main?
PS: Cool that you continue to work on this - I have nothing else ongoing, so let's do this :)

**Comment by thrillfall** [comment]
> @thrillfall I will take a look tomorrow. Just the usual request: can you rebase everything onto main?
> PS: Cool that you continue to work on this - I have nothing else ongoing, so let's do this :)

Yes. Its fun to do this. So thank you for the opportunity.
As far as i can see this branch is rebased to main HEAD. Or is there another main?

**Comment by dennisguse** [comment]
PS: I did not check if this PR was rebased onto main - all fine then :)
PPS: having fun is cool ><

File: src/main/java/de/dennisguse/opentracks/io/file/exporter/InstantPostWorkoutExport.java
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/io/file/exporter/InstantPostWorkoutExport.java:31]: TrackExportThread can be replaced by `new Thread(() -> trackFileFormat.newTrackExporter(context).writeTrack(track, outputStream)).start();`

File: src/main/java/de/dennisguse/opentracks/io/file/exporter/InstantPostWorkoutExport.java
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/io/file/exporter/InstantPostWorkoutExport.java:24]: Why this method name?
It is neither exporting the last track (trackId is passed) nor is it necessarily external storage: we just write into document uri: could be anything.

File: src/main/java/de/dennisguse/opentracks/io/file/exporter/InstantPostWorkoutExport.java
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/io/file/exporter/InstantPostWorkoutExport.java:28]: Method call can be improved by passing trackFileFormat to method and use it there.

File: src/main/java/de/dennisguse/opentracks/io/file/exporter/TrackExportThread.java
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/io/file/exporter/TrackExportThread.java:12]: Can be removed (see other comment)

File: src/main/java/de/dennisguse/opentracks/io/file/importer/SingleTrackExportDirectoryChooserActivity.java
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/io/file/importer/SingleTrackExportDirectoryChooserActivity.java:16]: Naming can be improved.

**Review by dennisguse — CHANGES_REQUESTED** [review]
Things to do:
* how to unset the "export track directory"? (and rename the label to "Default export directory")
* app crashes if export directory is not set and export all triggered (see stacktrace)
* auto export functionality cannot be disabled (we need a preference for this)
* do not use PreferencesUtils.getString / setString, but rather use PreferenceUtils.get/setYourPreference()
* do not hardcode preference ids in code, but define proper R.string (similar to all other preferences)
* do not hardcode translation "not yet set" in code, but use R.string
* update activities.txt with proper flow
* (minor) autoformat code
* can we rename this functionality to `autoExport`?
  Also the preferences are not named consistently (for my feeling); I know this is hard, but important to keep the code nice to work with everybody in the future. This also includes the activities (e.g., SingleTrackExportDirectoryChooserActivity) and helper functions (e.g., InstantPostWorkoutExport).
  Consistent naming is key for long-term happiness ;)

Question: do we need some user feedback while the track is exported?

I am not a big fan, of multi-line statements (for me it is easier to read if the statement is in one line).
If it get's to big it should usually be split into two lines anyhow.

PS: Code looks good.

PPS: I still would recommend to squash the commits.
Usually, I would go through each commit and review them individually.
But the 22 commits to not tell a story, but are changes at different places that sometimes get changed again.
So, I only read all commits together (limiting the value of individual commits).
Keeping them separate only adds value if the changes are useful to consider in an isolated fashion.

```
     Caused by: java.lang.NullPointerException: uriString
        at android.net.Uri$StringUri.<init>(Uri.java:496)
        at android.net.Uri$StringUri.<init>(Uri.java:486)
        at android.net.Uri.parse(Uri.java:458)
        at de.dennisguse.opentracks.util.PreferencesUtils.getInstantExportDirectoryUri(PreferencesUtils.java:381)
        at de.dennisguse.opentracks.io.file.InitiateExportActivity.onCreate(InitiateExportActivity.java:18)
        at android.app.Activity.performCreate(Activity.java:7824)
        at android.app.Activity.performCreate(Activity.java:7813)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1306)
```

**Comment by thrillfall** [comment]
Thanks for the review. I'll fix all but could take some time (since i currently don't have so much).a

Auto export functionality can be disabled. There is a boolean switch setting for it. Or doesn't it work?

**Comment by thrillfall** [comment]
how to unset the "export track directory" ?

**Comment by dennisguse** [comment]
About unset: I don't know. I just know that we need this ;)
About Boolean switch: it is not shown in SettingsActivity.

And take your time, we are in no hurry.

**Comment by dennisguse** [comment]
@pstorch @rgmf Do you any suggestion on how to unset a set export directory?
I can think of the following (order of no relevance):
* long-press the preference (optional: open an unset confirmation dialog)
* after opening the directory chooser, the user cancels the process (e.g., back button)
* add another "reset export directory" preference

**Comment by rgmf** [comment]
> @pstorch @rgmf Do you any suggestion on how to unset a set export directory?
> I can think of the following (order of no relevance):
> 
>     * long-press the preference (optional: open an unset confirmation dialog)
> 
>     * after opening the directory chooser, the user cancels the process (e.g., back button)
> 
>     * add another "reset export directory" preference

In NextCloud there is something similar when you set an auto-upload folder. To unset it there is a dialog with "Delete", "Cancel" and "Save" action buttons.

**Comment by pstorch** [comment]
Either e long click on the export directory preference or a "trashcan" icon next to it.

File: src/main/java/de/dennisguse/opentracks/io/file/exporter/InstantPostWorkoutExport.java
Comment (thrillfall) [reply src/main/java/de/dennisguse/opentracks/io/file/exporter/InstantPostWorkoutExport.java:31]: If i do it like you proposed it will only export 37bytes without any exception. The file does not contain all the data.

**Comment by dennisguse** [comment]
@thrillfall I marked this PR as draft; when you are ready, just change it back.
PS: If you need help or something else, just let me know.

File: src/main/java/de/dennisguse/opentracks/settings/SettingsActivity.java
Comment (thrillfall) [reply src/main/java/de/dennisguse/opentracks/settings/SettingsActivity.java]: if i'd use null i'd have to set the fallback value for the summary explicitly. This way its automagically showing "not set yet" in the settings activity

File: src/main/java/de/dennisguse/opentracks/io/file/exporter/TrackExportThread.java
Comment (thrillfall) [reply src/main/java/de/dennisguse/opentracks/io/file/exporter/TrackExportThread.java:12]: it does not work the way you suggested (see other comment)

**Comment by thrillfall** [comment]
I resolved everything but the Thread as closure approach (does not work as mentioned) and offering to unset the default export directory.
Problem: If the user does not want to use the default directory (i assume this is ain't rare) the "unsetting approach" would force the user to unset the default directory AND then export AND then reset the default directory everytime. 

I thought about it and came to the conclusion it would be better if we would ask the user everytime ze clicks "export all" if ze wants to export to "default directory" or a custom one (a file dialog would open then).
What do you think?

**Comment by dennisguse** [comment]
About the described unsettings problem: I agree that might be a annoying, but I am not sure how often this will really take place.
I guess that in 99.9% of the time, users will just use the default export directory and that's it.
I believe we can worry about this if somebody complains ;)
So, I would stick with the current solution, i.e., if the export directory is set, "export all" should export there.

About unsetting in general: what about using an intermediate dialog that shows: cancel, set, unset as suggested by @rgmf here: https://github.com/OpenTracksApp/OpenTracks/pull/383#issuecomment-694686041

**Comment by thrillfall** [comment]
And there we go assuming what 99% of the users are going to to and how with the app. And we both have different views on what our 99% are going to do ;-)

From my point of view its like that: I use the default directory for instant export. And it won't change. If I export all I export to a different directory (because the instant exports are already in the default directory). And selecting a custom directory for that single export all should not be done by unsetting the default. I have the default set for all the recurring usages (like instant export). The rare usage cases (like export all) should ask if one wants to use the default.

Let's see how often we need to exchange our personal expectations on how the user journey looks like until I dont have to implement an ugly save/set/unset-dialog :-)
Which by the way does not exist in the nextcloud app

**Comment by thrillfall** [comment]
And to be frank. I never wanted to change the ”export all" feature because the situation we now have. We discuss UX of things that are out of scope of post-workout-export.
But I gave you the benefit of being head maintainer and did what you asked for (have export-all benefit from a default export directory). But this broad discussion on how to unset the default where the necessity of unsetting it only originates from a functionality that was never to be changed is somewhat unnerving.

**Comment by thrillfall** [comment]
I have added the dialog for "export all" to choose between the default directory or selection a custom one. If no default directory was set "export all" behaves like before
Screenshot_1601983559

**Comment by thrillfall** [comment]
As far as i am concerned all change requests have been addressed and are resolved. 
The only unresolved request is the lambda function for the thread. I leave that to more skilled people to improve.

**Comment by dennisguse** [comment]
@thrillfall Can you revert/remove https://github.com/OpenTracksApp/OpenTracks/pull/383/commits/78c5072ef61de5d4108a6b8f2bd47e9020135637? It works, but it has some flaws (and it I guess it should have been a fragment rather than an activity, but I also don't know how to do this ><)
Then just squash the commits (into some meaningful units, can also be 1) and I will take care at the weekend.
I will then take care at the weekend by adding a simple approach to unset the

**Comment by thrillfall** [comment]
> @thrillfall Can you revert/remove [78c5072](https://github.com/OpenTracksApp/OpenTracks/commit/78c5072ef61de5d4108a6b8f2bd47e9020135637)? It works, but it has some flaws (and it I guess it should have been a fragment rather than an activity, but I also don't know how to do this ><)
> Then just squash the commits (into some meaningful units, can also be 1) and I will take care at the weekend.
> I will then take care at the weekend by adding a simple approach to unset the

sure thing. enjoy the code ;-)

**Comment by pstorch** [comment]
I did a little testdrive of this feature. I'll want to enable it when it is released, too.

My suggestion to the directory (un-)selection discussion:
I would split "Export all" and "Instant post-workout export". For "Export all" I would liked to be asked all the time for the destination directory. Because this is a manual action and I might want to choose different location at the moment.
For "Instant post-workout export" we need to keep the selected target directory because it is automatic. If the user doesn't want to use it anymore he can disable the feature. The last chosen directory can still be kept.

BTW: the "Export all" function exports all tracks over and over again. I would expect a "This track already exists at the destination directory! [overwrite], [skip], [ ] for all"
@dennisguse should I open an issue for this?

**Comment by pstorch** [comment]
When the configured export directory is gone, there is no feedback to the user.

**Comment by thrillfall** [comment]
@pstorch This behavior is exactly what I implemented. But   I was asked to revert it in favour of unsetting the default directory.
I am glad that you like the feature

**Comment by dennisguse** [comment]
@pstorch @thrillfall My intention is to merge this quickly as this PR took quite some time - so I will just hack something and merge it.
Then we can move follow-up work to a new PR.

About the commit (I asked to revert), it should have been using a fragment rather than an activity; it also had some other issues (like order of buttons etc). And it was not really necessary for the core feature.

**Comment by thrillfall** [comment]
> @pstorch @thrillfall My intention is to merge this quickly as this PR took quite some time - so I will just hack something and merge it.
> Then we can move follow-up work to a new PR
> 
> About the commit (I asked to revert), it should have been using a fragment rather than an activity; it also had some other issues (like order of buttons etc). And it was not really necessary for the core feature.

The core feature (instant export) was ready a month ago (after fixing raised issues) but was "enriched" with wishes to "integrate this nicely with export all". I said this further integration should go into another PR because it was not necessary for the core feature and would just hold up merging this by piling up addition after addition following issues because of those. And there we took another iteration to discuss how to deal with follow up issues like unsetting the default directory.
That is why this PR "took quite some time". I addressed any additional integration that was wished for to get the core feature into the app for more then 2 month.
But suddenly now its due and time to "hack something and merge it"?
I feel really stupid.

**Comment by dennisguse** [comment]
What happens next: I will spend some time checking the code, refactoring it (if necessary), and add the unset directory functionality (if the DirectoryChooserActivity does not return SUCCESS).
Then let's revisit the functionality and make it even cooler.

@thrillfall Thanks for the feedback and I agree: it could have been a more pleasant thing if I would have handled things differently.
For me, some lessons learned for the next time.

Long story short: want to join us?
Then you can complain about others messing around with the code - it is whole different experience ;)

**Comment by thrillfall** [comment]
@dennisguse I appreciate your words. But given your "what happens next" I doubt you learned what I hoped you would.
I don't want to "join". Half of my professional work is reviewing coworkers code (the other is coding myself). So I know the experience.
And what I learned is you repell and demotivate people to contribute if you change their contributions (not talking about coding issues here) by making demands on how you would like the contribution to work. And you don't motivate contributions if you say you'll refactor it after the code had several iterations of review and resolve. You don't motivate contributions if you "take over" the contributed code by changing behavior to your liking.
You don't motivate contributions if you (after 2 month hold up by requesting features outside of the contribution) say "let's revisit and make it even cooler".
The contribution was "cool" before unnecessarily mixing it up with "export all". And it even was cool when " export all" could use the new default directory.
And still no word on when you deem this contribution getting merged. Which would be the only counting act that says: I appreciate your contribution to the project.
Holding your fellow contributors in the hamster wheel for 2 month (i remember repeatedly opening new merge request with the same code) just says: your contribution will be appreciated (read: get merged) only whence It is exactly like I want it.
And from my professional experience this approach does not build team or encourage contributions from outside of that team.

I come to the conclusion that you don't mean to repel contributions because your words certainly don't. Yet part of your actions have done exactly that.

I closed the PR because I did not want to change anything anymore because the feeling that it would never be considered ready (not talking about bugs) was just too demotivating. I just wanted to leave it with you and let you do what ever you deem necessary.,

I use the instant export feature for 3 month now. Maybe in the future every user of the app can make use of it too.

I thank you all for your time (reviewing and contributing) and for your efforts on this great app.

I need to take a break from this PR. Its yours (and by the rule of open source always was) to do what ever you want with it. Close it, decide to make another app of it (like the graveyard of the statistics graph contribution) or just leave it here as a gravestone to make by passers aware of the struggles of open source software development ;-)

**Comment by dennisguse** [comment]
TL;DR: You are right. I will take care at the weekend.

@thrillfall 
I don't know how the code looks like (as it had so many changes, so I have to read and understand it again).
That's why I can't tell if I see the need to change something or if everything is beautifully build and just fits.
What I meant: I will make and will make changes without discussing them (before merging) as it just takes too much time at the moment.
I hate to do that, but I have to cut your and my loses here.
Nobody is happy with the situation.
It was never my intention to change any incoming PR as this is not how open-source should work (from my perspective) and it is demotivating.
But there is an exception to almost everything I guess.

From my perspective, you picked a topic that sounds very simple and then it exploded in terms of complexity.
The underlying issue is that you touched large parts of the code base that already before need some heavy improvements (e.g., TrackRecordingService and also the whole export to file code).
And I should have spent more time in the beginning to help you understand what parts should be touched (and how) rather than giving you code reviews afterwards.
The code base was not ready for this feature.
The whole import/export is till now far too error prone, lacks tests, and results in inconsistent behavior.
There are so many open issues in our tracker (e.g., inconsistent statistics on import; failure to import).
So, opening an issue for discussion on how it could have been done beforehand would have been helpful (I guess).

If you have time to contribute, really feel free to so.
I learned some lessons.

**Comment by dennisguse** [comment]
Closed in favor of #454.

**Pull request closed without merging**
